merkle_tree_wasm/src/merkle_tree_wasm.rs

```diff
@@ -0,0 +1,213 @@
+use ark_ff::PrimeField;
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use poseidon::{Poseidon, PoseidonConstants};
+use rayon::{iter::ParallelIterator, slice::ParallelSlice};
+
+pub use ark_ff;
+pub use ark_secp256k1;
+pub use num_bigint;
+pub use poseidon;
+
+pub struct MerkleTree<F: PrimeField, const WIDTH: usize> {
+    _marker: std::marker::PhantomData<F>,
+    leaves: Vec<F>,
+    poseidon: Poseidon<F, WIDTH>,
+    is_tree_ready: bool,
+    layers: Vec<Vec<F>>,
+    depth: Option<usize>,
+    pub root: Option<F>,
+}
+
+#[derive(CanonicalSerialize, CanonicalDeserialize)]
+pub struct MerkleProofBytes {
+    pub siblings: Vec<u8>,
+    pub path_indices: Vec<u8>,
+    pub root: Vec<u8>,
+}
+
+#[derive(Debug)]
+pub struct MerkleProof<F: PrimeField> {
+    pub leaf: F,
+    pub siblings: Vec<F>,
+    pub path_indices: Vec<usize>,
+    pub root: F,
+}
+
+impl<F: PrimeField, const WIDTH: usize> MerkleTree<F, WIDTH> {
+    pub fn new(constants: PoseidonConstants<F>) -> Self {
+        let arity = WIDTH - 1;
+
+        assert_eq!(arity, 2, "MerkleTree: Only arity 2 is supported.");
+
+        let mut poseidon = Poseidon::new(constants);
+        poseidon.state[0] = F::from(3u32);
+
+        Self {
+            poseidon,
+            leaves: Vec::new(),
+            _marker: std::marker::PhantomData,
+            is_tree_ready: false,
+            layers: Vec::new(),
+            depth: None,
+            root: None,
+        }
+    }
+
+    pub fn insert(&mut self, leaf: F) {
+        // Add the leaf to the bottom-most layer
+        self.leaves.push(leaf);
+    }
+
+    fn hash(poseidon: &mut Poseidon<F, WIDTH>, nodes: &[F]) -> F {
+        assert_eq!(nodes.len(), poseidon.state.len() - 1);
+
+        for i in 0..nodes.len() {
+            poseidon.state[i + 1] = nodes[i];
+        }
```

Comment:
```suggestion
        poseidon.state[1..(nodes.len() + 1)].copy_from_slice(nodes);
```

see 
https://rust-lang.github.io/rust-clippy/master/index.html#/manual_memcpy
https://rust-lang.github.io/rust-clippy/master/index.html#/redundant_slicing

(We should include `cargo clippy` and `cargo fmt` into repo wide quality checks in CI..., I'll open an issue for this)

Reply:
Good point, have you already opened an issue for that?

Reply:
https://github.com/anonklub/anonklub/issues/283